File: test/jte-runtime-cp-test-models/src/test/java/gg/jte/TemplateEngineTest.java
package gg.jte;

import gg.jte.generated.precompiled.Templates;
import gg.jte.output.StringOutput;
import gg.jte.runtime.TemplateUtils;
import org.assertj.core.api.AbstractThrowableAssert;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import test.Dummy;
import test.Model;

import java.lang.annotation.Annotation;
import java.util.Arrays;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;

/**
 * To run these tests, you first need to run the jte-maven-plugin (jte:precompile)
 */
public class TemplateEngineTest {

    private static TemplateEngine templateEngine = TemplateEngine.createPrecompiled(ContentType.Html);

    private static final gg.jte.generated.precompiled.Templates staticTemplates = new gg.jte.generated.precompiled.StaticTemplates();
    private static final gg.jte.generated.precompiled.Templates dynamicTemplates = new gg.jte.generated.precompiled.DynamicTemplates(templateEngine);
    Model model = new Model();

    public static Stream<Arguments> templates() {
        return Stream.of(
                Arguments.of(staticTemplates),
                Arguments.of(dynamicTemplates)
        );
    }

    @BeforeEach
    void setUp() {
        model.hello = "Hello";
        model.x = 42;
    }

    @ParameterizedTest
    @MethodSource("templates")
    void helloWorld(gg.jte.generated.precompiled.Templates templates) {
        String output = templates.helloWorld(model).render();
        assertThat(output).isEqualTo("Hello World");
    }

    @ParameterizedTest
    @MethodSource("templates")
    void exceptionLineNumber1(gg.jte.generated.precompiled.Templates templates) {
        Throwable cause = catchThrowable(() -> templates.exceptionLineNumber1(model).render());
        assertThat(cause).isInstanceOf(TemplateException.class)
            .hasCauseInstanceOf(NullPointerException.class)
                .hasMessage("Failed to render exceptionLineNumber1.jte, error at exceptionLineNumber1.jte:5");
    }

    @ParameterizedTest
    @MethodSource("templates")
    void unusedTag(gg.jte.generated.precompiled.Templates templates) {
        String output = templates.tagUnused("One", "Two").render();
        assertThat(output).isEqualTo("One is One, two is Two.");
    }
    @ParameterizedTest
    @MethodSource("templates")
    void excludedTemplates(gg.jte.generated.precompiled.Templates templates){
        assertThat(templates.getClass().getMethods()).noneMatch(m -> m.getName().contains("Exclude"));
    }

    @ParameterizedTest
    @MethodSource("templates")
    void normalContent(Templates templates) {
        String output = templates.main().render();
        assertThat(output).containsIgnoringWhitespaces("Header", "Main", "Footer");
    }

    @ParameterizedTest
    @MethodSource("templates")
    void nestedContent(Templates templates) {
        String output = templates.layout(templates.helloWorld(model)).render();
        assertThat(output).containsIgnoringWhitespaces("Header", "Hello World", "Footer");
    }


    @ParameterizedTest
    @MethodSource("templates")
    void hasAnnotation(Templates templates) {
        Class<? extends Templates> clazz = templates.getClass();
        Annotation[] annotations = clazz.getDeclaredAnnotations();
        assertThat(annotations).anyMatch(a -> a instanceof Dummy);
    }
}

File: test/jte-runtime-cp-test-models-gradle-kotlin/src/test/kotlin/gg/jte/TemplateEngineTest.kt
package gg.jte

import gg.jte.generated.precompiled.Templates
import gg.jte.generated.precompiled.DynamicTemplates
import gg.jte.generated.precompiled.StaticTemplates

import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

import org.assertj.core.api.Assertions
import org.assertj.core.api.ThrowableAssert

import java.util.stream.Stream
import java.util.function.Predicate

/**
 * To run these tests, you first need to run the gradle precompile task:
 */
class TemplateEngineTest {
    val model: test.Model = test.Model()

    @ParameterizedTest
    @MethodSource("templates")
    fun helloWorld(templates: Templates) {
        val output: String = templates.helloWorld(model).render()
        Assertions.assertThat(output).isEqualTo("Hello World")
    }

    @ParameterizedTest
    @MethodSource("templates")
    fun exceptionLineNumber1(templates: Templates) {
        val cause: Throwable =
            Assertions.catchThrowable(ThrowableAssert.ThrowingCallable {
                templates.exceptionLineNumber1(
                    model
                ).render()
            })
        Assertions.assertThat<Throwable>(cause).isInstanceOf(TemplateException::class.java)
            .hasCauseInstanceOf(NullPointerException::class.java)
            .hasMessage("Failed to render exceptionLineNumber1.kte, error at exceptionLineNumber1.kte:3")
    }

    @ParameterizedTest
    @MethodSource("templates")
    fun unusedTag(templates: Templates) {
        val output: String = templates.tagUnused("One", "Two").render()
        Assertions.assertThat(output).isEqualTo("One is One, two is Two.")
    }

    @ParameterizedTest
    @MethodSource("templates")
    fun normalContent(templates: Templates) {
        val output: String = templates.main().render()
        Assertions.assertThat(output).containsIgnoringWhitespaces("Header", "Main", "Footer")
    }

    @ParameterizedTest
    @MethodSource("templates")
    fun excludedTemplates(templates: Templates) {
        Assertions.assertThat(templates::class.java.getMethods()).noneMatch { m -> m.getName().contains("Exclude") }
    }

    @ParameterizedTest
    @MethodSource("templates")
    fun nestedContent(templates: Templates) {
        val output: String = templates.layout(templates.helloWorld(model)).render()
        Assertions.assertThat(output)
            .containsIgnoringWhitespaces("Header", "Hello World", "Footer")
    }

    @ParameterizedTest
    @MethodSource("templates")
    fun hasAnnotation(templates: Templates) {
        val clazz: Class<out Templates?> = templates::class.java
        val annotations: Array<Annotation> = clazz.getDeclaredAnnotations()
        Assertions.assertThat<Annotation>(annotations)
            .anyMatch(Predicate<Annotation> { a: Annotation? -> a is test.Dummy })
    }

    companion object {
        private val templateEngine: TemplateEngine = TemplateEngine.createPrecompiled(ContentType.Html)

        private val staticTemplates: Templates = StaticTemplates()
        private val dynamicTemplates: Templates = DynamicTemplates(templateEngine)
        @JvmStatic
        fun templates(): Stream<Arguments> {
            return Stream.of(
                Arguments.of(staticTemplates),
                Arguments.of(dynamicTemplates)
            )
        }
    }
}